src/torchcodec/samplers/_implem.py

```diff
@@ -0,0 +1,138 @@
+import random
+from typing import List, Optional
+
+import torch
+
+from torchcodec.decoders import FrameBatch, SimpleVideoDecoder
+
+
+def _validate_params(
+    *, decoder, num_clips, num_frames_per_clip, num_indices_between_frames
+):
+    if len(decoder) < 1:
+        raise ValueError(
+            f"Decoder must have at least one frame, found {len(decoder)} frames."
+        )
+
+    if num_clips <= 0:
+        raise ValueError(f"num_clips ({num_clips}) must be strictly positive")
+    if num_frames_per_clip <= 0:
+        raise ValueError(
+            f"num_frames_per_clip ({num_frames_per_clip}) must be strictly positive"
+        )
+    if num_indices_between_frames <= 0:
+        raise ValueError(
+            f"num_indices_between_frames ({num_indices_between_frames}) must be strictly positive"
+        )
+
+
+def _validate_sampling_range(
+    *, sampling_range_start, sampling_range_end, num_frames, clip_span
+):
+    if sampling_range_start < 0:
+        sampling_range_start = num_frames + sampling_range_start
+
+    # TODO: or max(sampling_range_start, num_frames - 1)?
+    sampling_range_start = sampling_range_start % num_frames
```

Comment:
Looking for input on this. I'm in a "let's try to avoid mid-train crashes as much as possible" mindset.

Reply:
why aren't we doing the same for sampling_range_end if it's not None?

Maybe you can extract out a function like _coerce_index() or something and apply it for both start and end?

Reply:
The question as I understand it is: what should the behavior be when the sampling start is greater than the total number of frames? I'm inclined to make this an error for now, for the following reasons:

1. None of the non-failure options (mod it or max it with number of frames) seem intuitive to me.
2. We already have a resilient way to do the right thing: _don't_ provide a start and end.
3. It's easy to relax what was once an error condition, but it's hard to change the behavior of non-error conditions.
4. This feels tied to the problem of what to do when there are not enough frames to meet the required frames per clip. We know we're going to do something there, but we're currently making it an error. When we have a mechanism to solve that problem, we could use it here, too.

Reply:
OK, let's just error for now and stay open for an automated behavior upon user request.

Reply:
> why aren't we doing the same for sampling_range_end if it's not None?

Good catch, thanks. I think we should cap it to `num_frames`, as this is a fairly common strategy for the upper bound (as in slice indexing).